sycl/include/sycl/ext/oneapi/experimental/user_defined_reductions.hpp

```diff
@@ -0,0 +1,148 @@
+//==-------------------- user_defined_reductions.hpp -----------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/defines.hpp>
+#include <sycl/group_algorithm.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext::oneapi::experimental {
+
+// ---- reduce_over_group
+template <typename GroupHelper, typename T, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<std::decay_t<GroupHelper>>), T>
+reduce_over_group(GroupHelper g, T x, BinaryOperation binary_op) {
+#ifdef __SYCL_DEVICE_ONLY__
+  T *Memory = reinterpret_cast<T *>(g.get_memory().data());
+  Memory[g.get_group().get_local_linear_id()] = x;
+  group_barrier(g.get_group());
+  T result;
+  if (g.get_group().leader()) {
+    for (int i = 0; i < g.get_group().get_local_linear_range(); i++) {
+      result = binary_op(result, Memory[i]);
+    }
+  }
+  group_barrier(g.get_group());
+  return result;
+#else
+  (void)g;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+template <typename GroupHelper, typename V, typename T,
+          typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<std::decay_t<GroupHelper>>), T>
+reduce_over_group(GroupHelper g, V x, T init, BinaryOperation binary_op) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return binary_op(init, reduce_over_group(g, x, binary_op));
+#else
+  (void)g;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+// ---- four reduce_over_group overloads with native binary_op
+template <typename GroupHelper, typename T, class BinaryOperation>
+sycl::detail::enable_if_t<
+    (is_group_helper_v<std::decay_t<GroupHelper>> &&
+     sycl::detail::is_complex<T>::value &&
+     sycl::detail::is_native_op<T, sycl::plus<T>>::value &&
+     sycl::detail::is_plus<T, BinaryOperation>::value),
```

Comment:
This is a overload for function https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/group_algorithm.hpp#L202, the purpose is to redirect call from the function from user_defined_reductions.hpp to the more optimized function in group_algorithm.hpp in case user decided to use `group_helper`, not a `group`